pkg/resource/managed_reconciler.go

```diff
@@ -0,0 +1,438 @@
+/*
+Copyright 2019 The Crossplane Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package resource
+
+import (
+	"context"
+	"time"
+
+	"github.com/pkg/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	"github.com/crossplaneio/crossplane/pkg/apis/core/v1alpha1"
+	"github.com/crossplaneio/crossplane/pkg/logging"
+	"github.com/crossplaneio/crossplane/pkg/meta"
+	"github.com/crossplaneio/crossplane/pkg/util"
+)
+
+const (
+	managedControllerName   = "managedresource.crossplane.io"
+	managedFinalizerName    = "finalizer." + managedControllerName
+	managedReconcileTimeout = 1 * time.Minute
+
+	defaultManagedShortWait = 30 * time.Second
+	defaultManagedLongWait  = 1 * time.Minute
+)
+
+// Error strings.
+const (
+	errGetManaged    = "cannot get managed resource"
+	errPublishSecret = "cannot publish managed resource connection secret"
+)
+
+// ConnectionDetails created or updated during an operation on an external
+// resource, for example usernames, passwords, endpoints, ports, etc.
+type ConnectionDetails map[string][]byte
+
+// A ManagedPublisher publishes the supplied ConnectionDetails for the supplied
+// Managed resource. ManagedPublishers must handle the case in which the
+// supplied ConnectionDetails are empty.
+type ManagedPublisher interface {
+	Publish(ctx context.Context, mg Managed, c ConnectionDetails) error
+}
+
+// A ManagedCompleter completes the creation of the supplied Managed resource.
+// This often involves adding a finalizer to the object metadata.
+type ManagedCompleter interface {
```

Comment:
I like this suggestion - good call.

Reply:
I took a quick pass at updating the PR with this suggestion.

I tried implementing this exactly as we discussed here - having `Reconcile()` explicitly add the finalizer immediately _before_ creating the external resource, then having it explicitly remove the finalizer immediately _after_ finalizing the external resource.

I opted to stick with using interfaces primarily because adding and removing finalizers requires us to update object metadata, which means we would have to call `r.client.Update()`. Currently the `Reconcile` method itself only ever does three things to the managed resource under reconciliation:

* Asks some method to operate on it (e.g. `external.Create()`)
* Modifies its conditioned status by calling `managed.SetConditions()`
* Propagates its status to the API server by calling `r.client.Status().Update()`

I like this separation of concerns because it simplifies the mental model when testing the `Reconcile()` method. We can assume it only ever updates conditions and status; everything else is done by an interface that we can stub out.

So in summary we now have two methods we call to update the non-status-bits of the managed resource under reconciliation in the API server:

* `r.managed.Establish()`, which is called immediately before the external resource is created. In practice there is only one implementation of this method, which adds a finalizer.
* `r.managed.Finalize()`, which is called immediately after the external resource is finalized. In practice there is only one implementation of this method, which removes a finalizer.

I'm not a huge fans of the names I've chosen here. I considered simply naming the methods `AddFinalizer` and `RemoveFinalizer`, but there's already a method that removes finalizers in from resource claims in the resource claim `Reconcile` method, and it is named `Finalize`.

Reply:
I like the idea around separation of `ExternalClient` and managed resource operation. Thanks for the change.

I am not sure about the naming though. In claim reconciler, `Finalize` is used for unbind operation and possibly finalizer deletion. But here we know that there has to be finalizer deletion. And we got `external.Finalize` as well, which is completely different.

If we're going with `external.Delete` as discussed in https://github.com/crossplaneio/crossplane/pull/593#discussion_r306174348 , I think `r.managed.Finalize` makes sense, because we'd imply that we use `Finalize` word as the last place you can do things after actual delete/unbind/deprovision operation anywhere in our interfaces, pretty much like a hook. If we keep `external.Finalize`, then a different name makes more sense to me.

Reply:
> We would act depending on the return of the Finalize. It could be deleted, error, deletion in progress.

I think we need either a struct like `ExternalFinalize` / `ExternalDeletion` or a string that `Finalize` can return so that we know whether it's time to remove the finalizer. Right now, as far as `Finalize` doesn't return error, we go ahead and remove the finalizer. This would implicitly require that `deletion in progress` is an error state(which I'd argue) that `external.Finalize` must return for us not to remove finalizer immediately. I don't think we have to have that implicitness here.

Reply:
Feel free to experiment with better names for the interfaces that add and remove finalizers, but please keep in mind that the resource claim reconciler currently uses the `Finalize` name. We should make sure there is symmetry between the two reconcilers, but it's fine if that means renaming some of the types used by the resource claim reconciler.

We also discussed in our meeting today that we should actively observe that the external resource was deleted before removing the finalizer by calling `external.Observe` before calling `external.Delete`. I like this suggestion, but we should be conscious that:

* We don't want our controller to block waiting for the delete to complete. We'd want to instead requeue.
* The default `kubectl` behaviour is to block upon deletion until the finalizer is removed. We may not want `kubectl` to block for 30+ seconds if it takes 30+ seconds for the external resource to be deleted.

Reply:
@negz What is the effect of `/status` for CRD being a 1.15 beta feature?
https://kubernetes.io/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definitions/#subresources

How would this affect a 1.13 user?

Reply:
Since we landed on using `Delete` in external client both for name and functionality, I think `Establish` & `Finalize` couple makes sense. I see that in claim controller, this corresponds to `Bind` & `Finalize`. I mean one could argue the naming in that context as well but I don't think that would be in the scope of that PR.

> The default kubectl behaviour is to block upon deletion until the finalizer is removed. We may not want kubectl to block for 30+ seconds if it takes 30+ seconds for the external resource to be deleted.

I am not sure whether this is something _we_ should be worried about. See https://github.com/kubernetes/kubernetes/issues/66110 That's the design choice they made that I'd argue but I'm not sure if we should limit our functionality because of that. It will hang until the finalizer is removed by the controller, meaning it will take as long as the full deletion. However, user can cancel the call at any time and since the deletion timestamp is put on the very second you called `kubectl delete`, the deletion will continue.

Here is the commit for this change https://github.com/crossplaneio/crossplane/commit/e42043ef6b6d0776601fd416f0f079cace4c5daf